lib/versioner/util.go

```diff
@@ -38,11 +57,215 @@ func UntagFilename(path string) (string, string) {
 	versionTag := ExtractTag(path)
 
 	// Files tagged with old style tags cannot be untagged.
-	if versionTag == "" || strings.HasSuffix(ext, versionTag) {
+	if versionTag == "" {
 		return "", ""
 	}
 
+	// Old style tag
+	if strings.HasSuffix(ext, versionTag) {
+		return strings.TrimSuffix(path, "~"+versionTag), versionTag
+	}
+
 	withoutExt := path[:len(path)-len(ext)-len(versionTag)-1]
 	name := withoutExt + ext
 	return name, versionTag
 }
+
+func retrieveVersions(fileSystem fs.Filesystem) (map[string][]FileVersion, error) {
+	files := make(map[string][]FileVersion)
+
+	err := fileSystem.Walk(".", func(path string, f fs.FileInfo, err error) error {
+		// Skip root (which is ok to be a symlink)
+		if path == "." {
+			return nil
+		}
+
+		// Skip walking if we cannot walk...
+		if err != nil {
+			return err
+		}
+
+		// Ignore symlinks
+		if f.IsSymlink() {
+			return fs.SkipDir
+		}
+
+		// No records for directories
+		if f.IsDir() {
+			return nil
+		}
+
+		path = osutil.NormalizedFilename(path)
+
+		name, tag := UntagFilename(path)
+		// Something invalid, assume it's an untagged file
+		if name == "" || tag == "" {
+			versionTime := f.ModTime().Truncate(time.Second)
+			files[path] = append(files[path], FileVersion{
+				VersionTime: versionTime,
+				ModTime:     versionTime,
+				Size:        f.Size(),
+			})
+			return nil
+		}
+
+		versionTime, err := time.ParseInLocation(TimeFormat, tag, locationLocal)
+		if err != nil {
+			// Can't parse it, welp, continue
+			return nil
+		}
+
+		if err == nil {
+			files[name] = append(files[name], FileVersion{
+				VersionTime: versionTime.Truncate(time.Second),
+				ModTime:     f.ModTime().Truncate(time.Second),
+				Size:        f.Size(),
+			})
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return files, nil
+}
+
+type fileTagger func(string, string) string
+
+func archiveFile(srcFs, dstFs fs.Filesystem, filePath string, tagger fileTagger) error {
```

Comment:
I propose to replace `tagger fileTagger` with `noTag bool` (and remove `type fileTagger` in general. Follow up at https://github.com/syncthing/syncthing/pull/5514/files#diff-607726c1924296ce5c4bb168265bf986R175

Reply:
I don't see how this makes it more flexible?

Reply:
It makes it less flexible, which is exactly what I intend to fix what I was originally confused about: Archiving is flexible (it uses `fileTagger`) and restoring is not (it uses `Un`-/`TagFilename` which understands "normal" tag or no tag). So I propose to reduce the flexibility to just tag or no tag instead of `fileTagger`.

Reply:
I dunno, I kind of disagree with you.

Reply:
Ok fine with me.